### PR TITLE
Add support for form async validators + unique custom validator

### PR DESCRIPTION
**Pull request by stissot**
Async Form Validators are useful to validate a form field value using a backend API.

This adds `getFormAsyncValidators()` similar to `getFormValidators()`, called by `getFormConfig()`.

Is also adds the generic `uniqueValidator()` method returning an `AsyncValidatorFn` that can be used to check any model and field for existing values.

Usage in model (e.g `Product.code`):

```
    public getFormAsyncValidators(): FormAsyncValidators {
        return {
            code: [this.uniqueValidator('code')],
        };
    }
```

### CODE REVIEW

**Comment by sambaptista** [comment]
Validators are relative to forms, not to model or services and should be standalone. They should be in their specific class instead of NaturalAbstractModelService. 

Can you refactor into an NaturalValidators.unique(args). The count function may be out of the class context but in the same file without export mention. 

`NaturalValidators.ts`
```ts
// Not exported variable, not in class neither. It's like a private function for current file.
const count = (queryVariablesManager, name) => {
   ...
}

export class NaturalValidators {

    public static unique(fieldName: string, service: AbstractModelService<any, any, ...> , apollo: Apollo, name: string)   {
       ...
       count(...);
        ...
    }
}
```


Usage would be more standard 

```ts
public getFormAsyncValidators(): FormAsyncValidators {
        return {
            code: [NaturalValidators.unique('code', this.userService, this.apollo, 'user')],
            numberField : [Validators.min(10)] // angular native validator usage for comparison
        };
}
```

File: projects/natural/src/lib/services/abstract-model.service.ts [hunk 1]
```diff
@@ -108,6 +122,19 @@ export abstract class NaturalAbstractModelService<Tone,
...
+        for (const key of Object.keys(asyncValidators)) {
+            if (config[key] && asyncValidators[key]) {
+                config[key].setAsyncValidators(asyncValidators[key]);
+            } else {
+                const formState = {
+                    value: model[key] ? model[key] : null,
+                    disabled: disabled,
+                };
+
+                config[key] = new NaturalFormControl(formState, null, asyncValidators[key]);
```
Comment (sambaptista) [inline projects/natural/src/lib/services/abstract-model.service.ts:141]: I'm unsure about that "null". Don't change that for now, I'll test it by myself, and implement tests.

**Comment by sambaptista** [comment]
The rest seems ok, good job. 

When the refactor is done, we can merge. We have to keep in mind that a debounce is still to do (in another commit/PR)

File: projects/natural/src/lib/services/abstract-model.service.ts [hunk 2]
```diff
@@ -456,6 +483,48 @@ export abstract class NaturalAbstractModelService<Tone,
...
+            if (control.value) {
+                const qvm = new NaturalQueryVariablesManager<Vall>();
+                const condition = {};
+                condition[fieldName] = {equal: {value: control.value}};
+                const variables: any = {
+                    pagination: {pageIndex: 0, pageSize: 0},
+                    filter: {groups: [{conditions: [condition]}]},
+                };
+                qvm.set('variables', variables);
+                this.count(qvm).pipe(
```
Comment (sambaptista) [inline projects/natural/src/lib/services/abstract-model.service.ts]: Shouldnt this be returned ?

**Comment by stissot** [comment]
Thanks for the feedback. I've relocated the validator in its own class `NaturalValidators.unique()`, but left the `public count()` method in `NaturalAbstractModel`, because it needs to access properties (name, apollo) from the actual model to generate the query, and it can also be used to count objects matching the QueryVariables for other purposes than the validator. Also added some tests.

**Pull request merged**